samples/nrf9160/modem_shell/src/at/at_cmd_mode.c

```diff
@@ -0,0 +1,344 @@
+	at_cmd_len = 0;
+	k_mutex_unlock(&at_buf_mutex);
+
+	/* Check for the presence of one printable non-whitespace character */
+	for (const char *c = at_buf;; c++) {
+		if (*c > ' ') {
+			break;
+		} else if (*c == '\0') {
+			return; /* Drop command, if it has no such character */
+		}
+	}
+
+	/* Scan AT command buffer for possible pipelined at commands */
+	at_cmd_mode_work.pipe_cnt = 0;
+
+	if (strchr(at_buf, '|')) {
+		struct at_cmd_mode_pipelining_info *pipes = at_cmd_mode_work.pipe_infos;
+		char *next_char;
+		char *tmp;
+		int pipelined_at_cmd_len;
+
+		next_char = at_buf;
+		tmp = strtok(next_char, "|");
+
+		while (tmp != NULL) {
+			if (at_cmd_mode_work.pipe_cnt >= AT_MAX_CMD_PIPELINED) {
+				printk("\nMax %d pipelined at commands supported, dropping: \"%s\"",
+				       AT_MAX_CMD_PIPELINED, tmp);
+			} else {
+				pipelined_at_cmd_len = strlen(tmp);
+				if (pipelined_at_cmd_len > AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN) {
+					printk("\nPipelined AT cmd max len is %d,"
+					       " too long %d bytes AT cmd dropped",
+					       AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN,
+					       pipelined_at_cmd_len);
+				} else {
+					pipes[at_cmd_mode_work.pipe_cnt].pipelined_cmd_start = tmp;
+					pipes[at_cmd_mode_work.pipe_cnt].pipelined_cmd_len =
+						pipelined_at_cmd_len;
+					at_cmd_mode_work.pipe_cnt++;
+				}
+			}
+			tmp = strtok(NULL, "|");
+		}
+	}
+
+	/* Echo a line feed */
+	printk("\n");
+
+	writing = false;
+
+	k_work_submit_to_queue(&mosh_common_work_q, &at_cmd_mode_work.work);
+}
+
+#define CHAR_CAN 0x18
+#define CHAR_DC1 0x11
+
+static void at_cmd_mode_bypass_cb(const struct shell *sh, uint8_t *recv, size_t len)
+{
+	static uint8_t tail;
+	bool escape = false;
+
+	/* Check if escape criteria is met. */
+	if (tail == CHAR_CAN && recv[0] == CHAR_DC1) {
+		escape = true;
+	} else {
+		for (int i = 0; i < (len - 1); i++) {
+			if (recv[i] == CHAR_CAN && recv[i + 1] == CHAR_DC1) {
+				escape = true;
+				break;
+			}
+		}
+	}
+
+	if (escape) {
+		/* Cannot use shell/mosh_print in bybass callback */
+		printk("===========================================================\n");
+		printk("MoSh AT command mode exited\n");
+
+		shell_set_bypass(sh, NULL);
+		at_cmd_mode_dont_print = false;
+		at_monitor_pause(mosh_at_cmd_mode_handler);
+		at_cmd_mode_active = false;
+		inside_quotes = false;
+		tail = 0;
+		at_cmd_len = 0;
+		return;
+	}
+
+	/* Store last byte for escape sequence detection */
+	tail = recv[len - 1];
+
+	/* Handle byte by byte. */
+	for (int i = 0; i < len; i++) {
+		at_cmd_mode_cmd_rx_handler(recv[i]);
+	}
+}
+
+int at_cmd_mode_start(const struct shell *sh)
+{
+	if (at_cmd_mode_active) {
+		printk("AT command mode already started\n");
+		return -EBUSY;
+	}
+
+	k_work_init(&at_cmd_mode_work.work, at_cmd_mode_worker);
+	at_cmd_mode_work.pipe_cnt = 0;
+
+	at_monitor_resume(mosh_at_cmd_mode_handler);
+
+	term_mode = CONFIG_MOSH_AT_CMD_MODE_TERMINATION;
+
+	at_cmd_mode_active = !at_cmd_mode_active;
```

Comment:
or simply set at_cmd_mode_active to true, easier to read. Also, do you need the following if?

Reply:
yeah, need to simplify

Reply:
Simplified. Thanks for the comments!